signer/src/blockchain_sync.rs

```diff
@@ -0,0 +1,336 @@
+            return Err(Error::MissingBitcoinBlock(tenure.anchor_block_hash));
+        };
+
+        // We use `get_tenure` to fetch all of the Nakamoto blocks for each
+        // tenure, walking backwards, which accepts a `StacksBlockId` as its
+        // parameter. As we're walking backwards, we grab the first block in
+        // the tenure and use its parent block ID for the next iteration, which
+        // should give us the previous tenure and all of its blocks.
+        let first_block = tenure
+            .blocks()
+            .first()
+            .ok_or(Error::EmptyTenure(next_tenure_tip))?;
+
+        next_tenure_tip = first_block.header.parent_block_id;
+
+        // We re-use the same code as the block observer to write the blocks to
+        // the database.
+        block_observer::write_stacks_blocks(
+            &ctx.get_storage_mut(),
+            &ctx.config().signer.deployer,
+            &[tenure],
+        )
+        .await?;
+
+        // If the block height of the anchor block is equal to the nakamoto
+        // activation height then we're done.
+        if bitcoin_block.block_height == nakamoto_activation_height {
+            tracing::info!("nakamoto activation height reached, stacks block-sync completed");
+            return Ok(());
+        }
+    }
+}
+
+/// Helper function to wait for the Stacks node to reach the Nakamoto activation
+/// height.
+#[tracing::instrument(skip_all, fields(%nakamoto_activation_height))]
+pub async fn wait_for_nakamoto_activation_height(
+    ctx: &impl Context,
+    nakamoto_activation_height: u64,
+) -> Result<(), Error> {
+    let mut term = ctx.get_termination_handle();
+    let stacks_client = ctx.get_stacks_client();
+
+    let mut interval = tokio::time::interval(Duration::from_secs(5));
+    loop {
+        if let Ok(node_info) = stacks_client.get_node_info().await {
+            if node_info.burn_block_height > nakamoto_activation_height {
+                tracing::info!(
+                    current_height = %node_info.burn_block_height,
+                    "stacks node has reached the nakamoto activation height"
+                );
+
+                break;
+            }
+
+            tracing::info!(
+                current = %node_info.burn_block_height,
+                target = %nakamoto_activation_height,
+                "waiting for stacks node to reach nakamoto activation height"
+            );
+        }
+
+        tokio::select! {
+            _ = term.wait_for_shutdown() => return Ok(()),
+            _ = interval.tick() => {},
+        }
+    }
+
+    Ok(())
+}
+
+/// Waits for the Stacks node to report that it is fully-synced. It does this
+/// by polling the node's `info` endpoint every 5 seconds until the response's
+/// `is_fully_synced` field is `true`.
+#[tracing::instrument(skip_all)]
+pub async fn wait_for_stacks_node_to_report_full_sync(ctx: &impl Context) -> Result<(), Error> {
```

Comment:
In general it's not a risk per-se, but if the Stacks node isn't synced then we know for a fact that the signer is going to spam a bunch of errors while either trying to play coordinator with outdated information (likely spamming the network with a bunch of stuff that will be rejected) or trying to validate things from signers which it doesn't have information for.

So this is more of a cheap insurance that the Stacks node is functioning and up-to-date before starting up and joining p2p.

I guess my counter-question would be why not ensure things are in a good state before starting? For a relatively short pause in operation, this shouldn't result in much of a pause (if any). If we join p2p and start signing sweeps but can't handle validating mints, wouldn't it be better to not do either? 🤷🏻‍♂️

Reply:
Yeah I gave the wrong impression here. I'm not saying we shouldn't wait, but I was thinking: _shouldn't we be waiting more often_? That's what I meant by 
> I mean, are these risks that are unique to booting the sBTC signer, or are these ever-present issues?

I was also thinking about whether we should be waiting _differently_. For example, the signer should probably be updating their database with bitcoin data, so that the backfill is faster when their Stacks node comes back online. And maybe they should vote on deposits. But yeah there are definitely things that they shouldn't do, like signing rounds and DKG (because of the key rotation contract call logic).